src/codegen/compile.ml

```diff
@@ -2735,29 +2738,46 @@ module Object = struct
     )
 
   (* Returns a pointer to the object field (possibly following the indirection) *)
-  let idx_hash env indirect =
+  let idx_hash env m indirect =
     if indirect
-    then Func.share_code2 env "obj_idx_ind" (("x", I32Type), ("hash", I32Type)) [I32Type] (fun env get_x get_hash ->
+    then
+      let name = Printf.sprintf "obj_idx_ind<%d>" m in
+      Func.share_code2 env name (("x", I32Type), ("hash", I32Type)) [I32Type] (fun env get_x get_hash ->
       get_x ^^ get_hash ^^
-      idx_hash_raw env ^^
+      idx_hash_raw env m ^^
       load_ptr ^^ compile_add_const (Int32.mul MutBox.field Heap.word_size)
     )
-    else idx_hash_raw env
+    else idx_hash_raw env m
 
   (* Determines whether the field is mutable (and thus needs an indirection) *)
   let is_mut_field env obj_type s =
     let _, fields = Type.as_obj_sub [s] obj_type in
     Type.is_mut (Type.lookup_val_field s fields)
 
+  (* Computes a lower bound for the linear index of a field in an object *)
+  let field_lower_bound env obj_type s =
+    let open Type in
+    let _, fields = as_obj_sub [s] obj_type in
+    let fields = List.filter (function {typ = Typ _; _} -> false | _ -> true) fields in
```

Comment:
In the IR the object's type can have _type fields_, which are erased from the memory layout of the object. To not mess up the positional index calculation we'll erase them. A test for this feature is forthcoming.

Reply:
Thanks for the explanation. I think ideally it would be good to have different views of records, one for type checking, one for code generation (and maybe others for other stuff, like interpretation) would be good. But that kind of thing will require some refactoring..

Reply:
Turns out that there is a pass on IR that already have removed the type fields when they hit codeine, so yes I'll remove this and assert only.